datafn/python/datafn/utils/date.py

```
@@ -0,0 +1,34 @@
+            value = value.replace(tzinfo=timezone.utc)
+        return int(value.timestamp() * 1000)
+    if isinstance(value, str):
+        try:
+            # Handle Z suffix for Python < 3.11
+            s = value.replace("Z", "+00:00") if value.endswith("Z") else value
+            dt = datetime.fromisoformat(s)
+            if dt.tzinfo is None:
+                dt = dt.replace(tzinfo=timezone.utc)
+            return int(dt.timestamp() * 1000)
+        except (ValueError, TypeError):
+            raise ValueError(f"Invalid date value: {value}")
```

Ruff (0.15.7)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)